kernel: generic: patch: 802.2+LLC - set transport_header offset

APCCV opened the pull request:
With the introduction of DSA, network_header and transport_header offsets in skb for 802.2+LLC packets are short by two bytes when pulled out of the backlog.  This issue is caused by DSA as it appears after applying the commmit that converts an architecture to DSA.

skb buffer is processed without issue through GRO (if active), netif_receive and br_handle_frame as none of it relies on the offsets. Once skb processing reaches __netif_receive_skb_one_core, code resets network_header offset but assumes transport_header is accurate (if set), leaving it untouched.

For 802.2+LLC+SNAP frames this causes an issue as llc_rcv assumes the offset as correct and advances transport_header offset to denote the SNAP header location. Once skb processing reaches snap_rcv for handling, SNAP header is misaligned by two, failing match with registered handlers, and packet is silently dropped.

Patch modifies llc_rcv to reset transport_header offset and apply the LLC header size increment, instead of applying the increment on existing value. This allows snap_rcv to match OUI:PID values properly and restores normal behaviour.
An additional benefit would be that any non SNAP 802.2+LLC processing will now receive an accurate transport_header offset, although this code and 802.2+LLC supported protocols are very mature so if no other issues have been noticed already, odds are none will.

Issue does not apply to for Ethernet II frames supporting IPv4/v6 traffic as that stack most probably resets transport_header before using it (if it does at all). Regardless, this patch does not affect Ethernet II frame processing as Ethernet II frames are never handled by llc_rcv().

Patch and comments set to netdev@vger.kernel.org for feedback.

Tested successfully:
- on ipq806x (C2600) and ath79 (DIR-835).
- on snapshot and (although redundant but just in case) 23.05.

Discussion here:
https://forum.openwrt.org/t/dsa-and-appletalk-multicast-issues/216494/102

Open issue here:
https://github.com/openwrt/openwrt/issues/17095

APCCV commented:
Patch required to address this issue:
https://github.com/openwrt/openwrt/issues/17095

APCCV commented:
If merged please backport to 24.10.

ptpt52 commented:
mark

Neustradamus commented:
@APCCV: Thanks for your work on this!

systemcrash commented:
NVM, here it is :D

hauke commented:
Please send it to the netdev mailing list as an RFC patch, you can generate it with `git format-patch --rfc`. Often they react better to such patches than to open mails. I am not familiar with the code and can not say much about the content itself.
Please also add `Cc: stable@vger.kernel.org` in the real patch afetr you got a comment on the rfc one to make sure it gets backported.

APCCV commented:
Not sure I did what you expected but there it goes.

APCCV commented:
We got comments and they seem positive. They had me at “approach is correct” or something like.

Given that, can we merge this as a hack patch while I figure out all the niceties for a properly formatted and documented patch request?

It should not cause any undue burden to patch maintenance. This part of kernel code hasn’t changed in ages. Offset is the same for 23.05 kernel, to put in perspective. (This patch is not needed in 23.05, just tested it there in case).

systemcrash commented:
https://lore.kernel.org/netdev/ef68689e-7e0b-4702-a762-d214c7d76e3b@gmail.com/

APCCV commented:
> https://lore.kernel.org/netdev/ef68689e-7e0b-4702-a762-d214c7d76e3b@gmail.com/

That’s my rant. 

Feedback further down here:
https://lore.kernel.org/netdev/1fae5ceb-e15a-479a-b876-0239a4cdfc27@redhat.com/

hauke commented:
I think Paolo is fine with the patch, it has some formal issues.
After fixing the formal issues you should send it as a real patch to the netdev mailling list.

APCCV commented:
I get the same message out of his note.
OK, if that is the next step before we progress here, so be it. I’ll get that “formal request” done.

hauke commented:
Please send it upstream and then add the patch to the pending folder in OpenWrt.

APCCV commented:
Reading through the art of preparing a patch... the fun piece that hit is that it needs a "Fixes" tag.
Great!
And here I was thinking I could give up on digging through code until I figured out what causes the offset to be 2 bytes off. Nope.
Blame on the line points to commit, b0e380b1d8a8e0aca215df97702f99815f05c094. As fun as it was to find, that just involves a rename from `skb->h.raw` to `skb->transport_header`.  Irrelevant. Makes sense... patch is modifying something that probably dates to kernel version 0.9.
I get it. Patch is not fixing the root cause of the issue. Fix is a hack much later in the process after DSA screwed this up. The purist in me agrees this is not the right thing to do. The pragmatist says I don't want this hard enough.
Not complaining, but with that hurdle, this is parked until I have the time and drive to figure out what commit this "fixes".

Commit:
kernel: generic: patch: 802.2+LLC - set transport_header offset

Conversion to DSA broke 802.2+LLC+SNAP packet processing. Frames
received by napi_complete_done with GRO and DSA have transport_header
set two bytes short, or pointing 2 bytes before network_header &
skb->data. As snap_rcv expects transport_header to point to SNAP
header (OID:PID) after LLC processing advances offset over LLC header
(llc_rcv & llc_fixup_skb), code doesn't find a match and packet is
dropped.

Image built at this commit operates properly:
  86dadeba48 - generic: add patch for GPON-ONU-34-20BI quirk
Image built at following commit exhibits the issue:
  337e36e0ef - ipq806x: convert each device to DSA implementation

As issue is LLC specific, to avoid impacting non-LLC traffic, and to
follow up on original assumption made on kernel commit fda55eca5a33
("net: introduce skb_transport_header_was_set()") stating "network
stacks usually reset the transport header anyway", llc_fixup_skb to
reset and advance the offset. llc_fixup_skb already assumes the LLC
header is at skb->data, and by definition SNAP header immediately
follows.

Signed-off-by: Antonio Pastor <antonio.pastor@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17220
Signed-off-by: Robert Marko <robimarko@gmail.com>

APCCV commented:
> Please send it upstream and then add the patch to the pending folder in OpenWrt.

Done.
https://lore.kernel.org/netdev/20241220142020.1131017-1-@gmail.com/#t

robimarko commented:
You did not CC the maintaintainers as well as the patch is missing the prefix as to what tree you are targeting, so its going to take a while before anybody looks at it

APCCV commented:
Okay.
Is there anything I can do about it or I should let it be?
Reply to myself? New thread? What would you suggest at this point?

APCCV commented:
> You did not CC the maintaintainers as well as the patch is missing the prefix 

Well, as I said at the beginning of this adventure, I am taking this as a learning opportunity. What better way to than messing things up.

Thank you (and everyone else) for feedback. This is ahead by a century from where it was when I first thought "heck, I could try fix this...".

APCCV commented:
Alright. I think it's back on rails.
https://lore.kernel.org/netdev/CANn89iLQq9Uf+HJAAi30UO01SPR0q2jr6FKxmewz9v=8XjAwXA@mail.gmail.com/

Commit updated, file in pending directory.

APCCV commented:
It got applied! Woohoo!

https://lore.kernel.org/netdev/173532783475.575952.12564839557235746896.git-patchwork-notify@kernel.org/

Neustradamus commented:
@APCCV: Congratulations!
- https://git.kernel.org/pub/scm/linux/kernel/git/netdev/net.git/commit/?id=a024e377efed

robimarko commented:
Thanks! Rebased on top of main and merged!

Neustradamus commented:
@robimarko: Thanks for merging too :)

APCCV commented:
Oh awesome! Thank you!

Can we backport it to 24.10 please? Do you need a separate PR for that?

On 2024-12-28 04:56, Robert Marko wrote:
>
> Thanks! Rebased on top of main and merged!
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/openwrt/openwrt/pull/17220#issuecomment-2564284657>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AWPXVPNVC6276ZXQKKIIKUD2HZYVFAVCNFSM6AAAAABTJVLJNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNRUGI4DINRVG4>.
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>
>
--------------yQE9BddwEr3tlz5vFlvaxtvG
Content-Type: text/html; charset=UTF-8
Content-Transfer-Encoding: 8bit

<!DOCTYPE html><html><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
  </head>
  <body>
    <p>Oh awesome! Thank you!<br>
    </p>
    <p>Can we backport it to 24.10 please? Do you need a separate PR for
      that?</p>
    <div class="moz-cite-prefix">On 2024-12-28 04:56, Robert Marko
      wrote:<br>
    </div>
    <blockquote type="cite" ***@***.***">
      
      <p dir="auto">Thanks! Rebased on top of main and merged!</p>
      <p style="font-size:small;-webkit-text-size-adjust:none;color:#666;">—<br>
        Reply to this email directly, <a href="https://github.com/openwrt/openwrt/pull/17220#issuecomment-2564284657" moz-do-not-send="true">view it on GitHub</a>, or <a href="https://github.com/notifications/unsubscribe-auth/AWPXVPNVC6276ZXQKKIIKUD2HZYVFAVCNFSM6AAAAABTJVLJNWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKNRUGI4DINRVG4" moz-do-not-send="true">unsubscribe</a>.<br>
        You are receiving this because you were mentioned.<img src="https://github.com/notifications/beacon/AWPXVPLKYOMLB7W7BC67G6L2HZYVFA5CNFSM6AAAAABTJVLJNWWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTUY27QPC.gif" height="1" width="1" alt="" moz-do-not-send="true"><span style="color: transparent; font-size: 0; display: none; visibility: hidden; overflow: hidden; opacity: 0; width: 0; height: 0; max-width: 0; max-height: 0; mso-hide: all">Message
          ID: <span>&lt;openwrt/openwrt/pull/17220/c2564284657</span><span>@</span><span>github</span><span>.</span><span>com&gt;</span></span></p>
      <script type="application/ld+json">[
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/openwrt/openwrt/pull/17220#issuecomment-2564284657",
"url": "https://github.com/openwrt/openwrt/pull/17220#issuecomment-2564284657",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]</script>
    </blockquote>
  </body>
</html>

--------------yQE9BddwEr3tlz5vFlvaxtvG--

robimarko commented:
I backported it to 24.10 after merging to main, so its already there

Neustradamus commented:
Merged commits:
- Main: https://github.com/openwrt/openwrt/commit/da7ab64f1f6d72597a1c8a8bb3b57409227bc635
- 24.10: https://github.com/openwrt/openwrt/commit/76a546a83a68459b88a14b3382a58740daf5bf00

APCCV commented:
@hauke sorry, I won't make pinging you a habit.
Now that the patch is been backported to stable kernels (including 6.6) next kernel bump will require taking it out of the OpenWrt tree.
Should do anything for that?

robimarko commented:
@APCCV No need, the patch will just get dropped once kernel is updated to next stable point release.
